Grid search (part 2 of 2) (#57)
aml4td/website: chapters/grid-search.qmd

Some different models and pre-processors can have this quality, including the glmnet model (@sec-penalized-logistic-regression), partial least squares (@sec-pls), principal component feature extraction (@sec-pca), and others. 

Unfortunately, irregular designs cannot exploit the submodel trick since they are not factorial in nature. A hybrid design could be used where a dense sequence of $s_1$ parameter values is created for the tuning parameter associated with submodels and a separate space-filling design of size $s_2$ for the other parameters. These two grids can be crossed so that the space-filling design is replicated for each value of the submodel parameter. This produces $s_1\times s_2$ grid points but, effectively, only $s_2$ models are fit. 

> For our larger grid, we created a similar regular grid of `r nrow(large_sfd)` points with the same expanded ranges of parameters. There were `r sqrt(nrow(large_sfd))` unique, evenly spaced values for both parameters. To evaluate this grid, we only need to train `r sqrt(nrow(large_sfd))` models. Compared to the analgous space-filling design, the regular grid was `r round(large_grid_time[3]/large_grid_reg_time[3], 1)` times faster to evaluate. 

Reviewer: ```suggestion
For our larger grid, we created a similar regular grid of `r nrow(large_sfd)` points with the same expanded ranges of parameters. There were `r sqrt(nrow(large_sfd))` unique, evenly spaced values for both parameters. To evaluate this grid, we only need to train `r sqrt(nrow(large_sfd))` models. Compared to the analogous space-filling design, the regular grid was `r round(large_grid_time[3]/large_grid_reg_time[3], 1)` times faster to evaluate. 
```

Additionally, if expensive preprocessing is used, a naive approach where each pipeline is processed in parallel might be counterproductive because we unnecessarily repeat the same preprocessing. For example, suppose we are tuning a supervised model along with UMAP preprocessing. By sending each of the $s$ candidates to a worker, the same UMAP training occurs for each set of candidates that correspond to the supervised model. Instead, a _conditional process_ can be used where we loop across the UMAP tuning parameter combinations and, for each, run another loop across the parameters associated with the supervised model. 

One other consideration is memory. If the data used to train and evaluate the model are copied for each worker, the number of workers can be restricted to fit within system memory. 

In any case, parallel processing can almost always improve the speed of model tuning. 

### Racing {#sec-racing}

> 

Author: I will eventually write summaries of other methods (mostly for deep learning) like hyperopt, spearmint, etc. Not a high priority right now.

```{r}
#| label: fig-racing-initial
#| echo: false
#| out-width: 70%
#| fig-align: center
#| fig-width: 6
#| fig-height: 4.25
#| fig-cap: The first interim analysis in racing using a larger space-filling design for the boosted tree model.

brier_param %>% 
  ggplot(aes(y = candidate)) + 
  geom_point(aes(x = estimate), cex = 1 / 2) +
  geom_vline(xintercept = 0, lty = 2) +
  geom_errorbar(aes(xmin = conf.low, xmax = estimate, col = Decision), 
                alpha = 1 / 3) +
  labs(x = "Drop in Brier Score", y = "Candidate Rank")
```

In the end, the racing process eliminated all but `r num_remaining` candidates. The process eliminated candidates at iterations `r loss_text[[1]]`. In all, `r num_fit` models were fit out of the possible `r format(nrow(large_sfd) * nrow(sim_rs), big.mark = ",")` (`r round(num_fit / (nrow(large_sfd) * nrow(sim_rs)) * 100, 1)`%). `r best_text`

> The model in @eq-perf-mod-racing allows us to estimate the within-resample correlation coefficient. This estimates how similar the performance metric values are to one another relative to values between resamples. The correlation estimate is `r signif(within_corr, 2)`, indicating a moderate similarity of metrics within a resample. 

Reviewer: ```suggestion
The model in @eq-perf-mod-racing allows us to estimate the within-resample correlation coefficient. This estimates how similar the performance metric values are to one another relative to values between resamples. In our example, the estimate of within resample correlation is `r signif(within_corr, 2)`, and is a good example of why we should use @eq-perf-mod-racing when culling parameter candidates. 
```

## Optimization Bias and Nested Resampling {#sec-nested-resampling}

```{r}
#| label: 1d-boost
#| include: false
# Objects from nested_res.RData
nest_grid_brier <- collect_metrics(nest_grid_res)
grid_brier <- show_best(grid_res, metric = "brier_class", n = 1)
opt_per_resample <- 
  collect_metrics(grid_res, summarize = FALSE) %>%
  slice_min(.estimate, by = id) 
multiples <- 
  opt_per_resample %>% 
  count(learn_rate) %>% 
  filter(n > 1) %>% 
  pluck("learn_rate") %>% 
  log10() %>% 
  signif(., digits = 3)
lr_lab <- cli::pluralize("There {?was/were} {length(multiples)} learning rate{?s} selected more than once: ") 
lr_lab <- paste0(lr_lab, knitr::combine_words(paste0("10<sup>", multiples, "</sup>")), ".")
```

In the last section, we found that the numerically best candidate was `r best_large`. The corresponding resampling estimate of the Brier score was `r signif(best_grid$mean, 3)` with a standard error of `r signif(best_grid$std_err, 3)`. If someone were to ask us how well the optimal boosted tree performs, we would probably give them these performance estimates. 

> However, there may be an issue in doing so. We are using grid search and resampling to find the best combination of parameters _and_ to estimate the corresponding performance. the problem is that we don't know how accurate or precise the estimate of the _optimal_ candidate may be. In some situations, this dual use of the model tuning process can introduce **optimization bias** where, to some degree, our performance statistics are optimistic. 

Reviewer: ```suggestion
However, there is an issue in doing so that may affect our decision for selecting the optimal tuning parameters. We are using grid search and resampling to find the best combination of parameters _and_ to estimate the corresponding performance. The problem is that we don't know how accurate or precise the estimate of the _optimal_ candidate may be. In some situations, this dual use of the model tuning process can introduce **optimization bias** where, to some degree, our performance statistics are optimistic. 
```

This issue has been studied extensively in the high-dimensional biology literature, where there could be dozens of samples but thousands of predictors. In this case, feature selection is of paramount importance. However, the process of selecting important predictors in this situation can be very difficult and often leads to unstable models or preprocessing methods that overfit the predictor set to the data. Discussions of these issues can be found in @varma2006bias, @boulesteix2009optimal, @bischl2012resampling, and @bernau2013correcting.

> In this section, we'll introduce a few methods to quantify and correct for optimization bias. These approaches are not specific to traditional grid search; they can be used with racing, iterative search, or any algorithm that we use to optimize a model. Some of the concepts can be difficult and, for this reason, we'll use a simplified grid search scenario. Let's optimize our boosted tree by fixing the number of boosting iterations to 500 and only optimize the learning rate. The tools that we are about to discuss are not limited to a single tuning parameter; this simplified task can aid us in explaining and visualizing the process^[Unfortunately, this is an example that is not prone to optimization bias. We'll see examples later on this website where we can more effectively use these tools.].

Reviewer: ```suggestion
In this section, we'll introduce a few methods to quantify and correct for optimization bias. These approaches are not specific to traditional grid search; they can be used with racing, iterative search, or any algorithm that we use to optimize a model. Some of the concepts can be difficult and, for this reason, we'll use a simplified grid search scenario. Let's optimize our boosted tree by fixing the number of boosting iterations to 500 and only optimize the learning rate. But the tools that we are about to discuss are not limited to a single tuning parameter. Simplifying the task in this way allos us to visualize and explain the process^[Unfortunately, this is an example that is not prone to optimization bias. We'll see examples later on this website where we can more effectively use these tools.].
```

Using the same 10-fold cross-validation scheme, @fig-1d-boost shows the results of a conventional grid search over the learning rate where 100 values of that parameter were evaluated. As usual, the line indicates the average of the Brier scores produced by the 10 assessment sets. The results of this process indicates that the smallest Brier score is achieved with a learning rate of 10<sup>`r signif(log10(grid_brier$learn_rate), 3)`</sup>. The Brier score that corresponds to this tuning parameter value is estimated to be `r signif(grid_brier$mean, 3)` with corresponding standard error of `r format(grid_brier$std_err, digits = 3, scientific = FALSE)`.

::: {#fig-1d-boost}

::: {.figure-content}

```{r}
#| label: fig-1d-boost
#| echo: false
#| fig-align: center
#| out-width: 70%
#| fig-width: 7
#| fig-height: 4

autoplot(grid_res) +
  labs(y = "Brier Score") +
  geom_vline(xintercept = grid_brier$learn_rate, lty = 3) +
  geom_rug(data = opt_per_resample, aes(x = learn_rate, y = NULL, col = id), sides = "b", show.legend = FALSE, linewidth = 2, alpha = 3 / 4)
```

:::

> Basic grid search for learning rate with 100 grid points. The dotted vertical line shows the optimal learning rate using the overal resampling estimates. The colored rug values show the learning rates that were optimal _within each resample_. `r lr_lab`

Reviewer: ```suggestion
Basic grid search for learning rate with 100 grid points. The dotted vertical line shows the optimal learning rate using the overall resampling estimates. The colored rug values show the learning rates that were optimal _within each resample_. `r lr_lab`
```

$$\widehat{bias} = \frac{1}{B}\sum_{j=1}^B\left[\hat{Q}_j^{opt} - \hat{Q}_j^*\right]$$
For this particular example, the bias is fairly small (`r format(debiased_grid$mean_bias, digits = 3, scientific = FALSE)` with standard error `r format(debiased_grid$std_err_bias, digits = 3, scientific = FALSE)`). To correct the conventional estimator, we add the bias; the Brier score is adjusted from `r signif(debiased_grid$mean, 3)` to `r signif(debiased_grid$adjusted, 3)`. A more complex analytical method can be found in @tibshirani2018excess.

> We can also estimate optimization bias with more complex resampling schemes. Recall that the issue is that we are overextending out conventional resampling scheme by doing too much with the same data (i.e., estimating overall performance assessment and the optimal performance value). Nested resampling [@varma2006bias;@boulesteix2009optimal] prevents this overuse by using two layers of resampling: the "inner resamples" are used to estimate the optimal candidate values, and the "outer resamples" estimate performance at those values. 

Reviewer: ```suggestion
We can also estimate optimization bias with more complex resampling schemes. Recall that the issue is that we are overextending the conventional resampling scheme by doing too much with the same data (i.e., estimating overall performance assessment and the optimal performance value). Nested resampling [@varma2006bias;@boulesteix2009optimal] prevents this overuse by using two layers of resampling: the "inner resamples" are used to estimate the optimal candidate values, and the "outer resamples" estimate performance at those values. 
```

The panel on the left shows boxplots of the Brier scores for the inner resamples and the corresponding outer resample estimates. The standard error of the outer resamples is very similar to the the level of noise in the conventional estimate.  

In the end, the nested resampling estimate of the Brier score was estimated as `r signif(nest_grid_brier$mean, 3)`; a value very close to the single 10-fold cross-validation result shown in @fig-1d-boost.

::: {.important-box}
It is important to emphasize that nested resampling is for **verification, not optimization**. It provides a better estimate of our model optimization process; it does not replace it. 
:::

We might best communicate the results of nested resampling like this: 

> > We tuned the learning rate of our boosted classification tree using grid search and found that a rate of 10<sup>`r signif(log10(grid_brier$learn_rate), 3)`</sup> to be best. We think that this model has a corresponding Brier score, measured without optimization bias, of `r signif(nest_grid_brier$mean, 3)`. 

Reviewer: Do you think it would be useful to add a confidence interval to this estimate?

> We can look at the candidates produced by the inner resamples to understand the stability of the optimization process and potentially diagnose other issues. We _would not_ choose "the best" inner resampling result and move forward with its candidate value. 

Reviewer: I think it would be good to explain why we shouldn't use the best inner resampling result.  But I can't quite come up with the explanation.

Author: I added a note on that.

It is a good idea to use nested resampling when the training set is small, the predictor set is very large, or both. The lack of significant bias in the analysis above does not discount the problem of optimization bias. It exists but can sometimes be within the experimental noise. 

It is a good idea to choose an inner resampling scheme that is as close as possible to the single resampling scheme that you used for optimization. The outer scheme can vary depending on your needs. Care should be used when the outer resampling method is the bootstrap. Since it replicates training set points in its analysis sets, the inner resamples need to use the _unique_ rows of the original training set. Otherwise the same data point might end up in both the inner analysis and assessment sets. 

The primary downside to nested resampling is the computational costs. Two layers of resampling have a quadratic cost, and only the inner resamples can be executed in parallel. Using `r num_workers` workers, the analysis in @fig-nested took `r sprintf("%2.1f", grid_nest_time[3]/grid_time[3])`-fold longer to compute than the basic grid search that produced @fig-1d-boost.

## Setting Parameter Ranges

> Specifying the range of parameters use to define the grid may involve some guesswork. For some parameter types, there is a well understood and defined range. For others, such as learning rate, there is a lower bound (zero) and a loosely defined upper bound mostly based on convention and prior experience. For grids, we want to avoid configurations with few unique values and a wide range. This might not sample the section of the parameter space that includes nonlinearity and the region of optimal results. @fig-learn-rate-seq shows an example of this for the learning rate. The initial grid included three points that missed the regions of interest. @fig-learn-rate-grid did a better job with more grid points. In high dimensional parameter space, the likelihood of a poor grid increases, especially for relatively “small” grid sizes. 

Reviewer: ```suggestion
Specifying the range of parameters used to define the grid may involve some guesswork. For some parameter types, there is a well understood and defined range. For others, such as learning rate, there is a lower bound (zero) and a loosely defined upper bound mostly based on convention and prior experience. For grids, we want to avoid configurations with few unique values and a wide range. This might not sample the section of the parameter space that includes nonlinearity and the region of optimal results. @fig-learn-rate-seq shows an example of this for the learning rate. The initial grid included three points that missed the regions of interest. @fig-learn-rate-grid did a better job with more grid points. In high dimensional parameter space, the likelihood of a poor grid increases, especially for relatively “small” grid sizes. 
```

> We often envision of the relationship between a predictor and model performance as being a shark peak. If we cannot find the pinnacle, the model optimization would fail. Luckily, as seen in @fig-nested, there are often substantial regions of parameter space with good performance. Model optimization may not be superficially easy but it is often not impossible. We only need to sample the optimal region once to be successful. 

Reviewer: ```suggestion
We often envision the relationship between a predictor and model performance as being a sharp peak. If we cannot find the pinnacle, the model optimization would fail. Luckily, as seen in @fig-nested, there are often substantial regions of parameter space with good performance. Model optimization may not be superficially easy but it is often not impossible. We only need to sample the optimal region once to be successful. 
```
